src/api/core/ciphers.rs

```
@@ -297,6 +298,46 @@ fn put_cipher(uuid: String, data: Json<CipherData>, headers: Headers, conn: DbCo
     Ok(Json(cipher.to_json(&headers.host, &headers.user.uuid, &conn)))
 }
 
+#[derive(Deserialize)]
+#[allow(non_snake_case)]
+struct CollectionsAdminData {
+    collectionIds: Vec<String>,
+}
+
+#[post("/ciphers/<uuid>/collections-admin", data = "<data>")]
+fn post_collections_admin(uuid: String, data: Json<CollectionsAdminData>, headers: Headers, conn: DbConn) -> EmptyResult {
+    let data: CollectionsAdminData = data.into_inner();
+
+    let cipher = match Cipher::find_by_uuid(&uuid, &conn) {
+        Some(cipher) => cipher,
+        None => err!("Cipher doesn't exist")
+    };
+
+    if !cipher.is_write_accessible_to_user(&headers.user.uuid, &conn) {
+        err!("Cipher is not write accessible")
+    }
+
+    let posted_collections: HashSet<String> = data.collectionIds.iter().cloned().collect();
+    let current_collections: HashSet<String> = cipher.get_collections(&conn).iter().cloned().collect();
+
+    //TODO: update cipher collection mapping
+    for collection in posted_collections.symmetric_difference(&current_collections) {
+        match Collection::find_by_uuid(&collection, &conn) {
+            None => (), // Does not exist, what now?
```

I think in this case we should err! with something like "Invalid collection ids provided", even though it's something that shouldn't happen. Same thing with the else case of the is_writable_by_user a couple of lines down.

---

My problem with err!-ing here is that this might happen after couple loops, so the changes for other collections could have been already applied. In other parts of the API (like the mass move) we ignore errors on individual operations and only error on common problems. (like cipher ownership in this case)

I  think it might make more sense to apply as much as we can and just skip whatever fails. I'm wondering how upstream deals with that, if there's some way to report partial success.

---

I take that back, I went to check the code in the other cases (`/ciphers/delete`, `/ciphers/move`) and we indeed fail mid-loop. I'll add some `err!` here as well. Apologies for confusion.